wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.mm

```diff
@@ -0,0 +1,129 @@
+                                            0);
+            NSArray* results = [self readImageView:imageView];
+            CVPixelBufferUnlockBaseAddress(pixelBuffer, kCVPixelBufferLock_ReadOnly);
+            return results;
+    }
+
+    // If given pixel format is not a supported type with a luminance channel we just use the
+    // default method
+    return [self readCIImage:[[CIImage alloc] initWithCVImageBuffer:pixelBuffer]];
+}
+
+- (NSArray<ZXIResult *> *)readCIImage:(nonnull CIImage *)image {
+    CGImageRef cgImage = [self.ciContext createCGImage:image fromRect:image.extent];
+    auto results = [self readCGImage:cgImage];
+    CGImageRelease(cgImage);
+    return results;
+}
+
+- (NSArray<ZXIResult *> *)readCGImage: (nonnull CGImageRef)image {
+    CGColorSpaceRef colorSpace = CGColorSpaceCreateWithName(kCGColorSpaceGenericGray);
+    CGFloat cols = CGImageGetWidth(image);
+    CGFloat rows = CGImageGetHeight(image);
+    NSMutableData *data = [NSMutableData dataWithLength: cols * rows];
+
+
+    CGContextRef contextRef = CGBitmapContextCreate(
+                                                    data.mutableBytes,// Pointer to backing data
+                                                    cols,                      // Width of bitmap
+                                                    rows,                     // Height of bitmap
+                                                    8,                          // Bits per component
+                                                    cols,              // Bytes per row
+                                                    colorSpace,                 // Colorspace
+                                                    kCGBitmapByteOrderDefault); // Bitmap info flags
+    CGContextDrawImage(contextRef, CGRectMake(0, 0, cols, rows), image);
+    CGContextRelease(contextRef);
+
+    ImageView imageView = ImageView(
+              static_cast<const uint8_t *>(data.bytes),
+              static_cast<int>(cols),
+              static_cast<int>(rows),
+              ImageFormat::Lum);
+    return [self readImageView:imageView];
+}
+
++ (DecodeHints)DecodeHintsFromZXIOptions:(ZXIDecodeHints*)hints {
+    BarcodeFormats formats;
+    for(NSNumber* flag in hints.formats) {
+        formats.setFlag(BarcodeFormatFromZXIFormat((ZXIFormat)flag.integerValue));
+    }
+    DecodeHints resultingHints = DecodeHints()
+        .setTryRotate(hints.tryRotate)
+        .setTryHarder(hints.tryHarder)
+        .setTryDownscale(hints.tryDownscale)
+        .setFormats(formats)
+        .setMaxNumberOfSymbols(hints.maxNumberOfSymbols);
+    return resultingHints;
+}
+
+- (NSArray<ZXIResult*> *)readImageView: (ImageView)imageView {
+    Results results = ReadBarcodes(imageView, [ZXIBarcodeReader DecodeHintsFromZXIOptions:self.hints]);
+
+    NSMutableArray* zxiResults = [NSMutableArray array];
+    for (auto result: results) {
+        if(result.error() == Error::None) {
```

Comment:
This does not compile. The way to go here is using `result.isValid()`.  But if you don't pass `DecodeHints::returnErrors(true)` and the default is not to, then all returned results are valid, so this check can completely be removed.

Also, the next line uses `.c_str()` which will fail if the `text()` property contains a `\0` byte, which may be the case.

Reply:
Thanks for pointing this out.

Regarding the `c_str()`. I'm afraid I don't quite understand. The function call actually expects a null terminated string. And doesn't a cstring contain a `\0`byte per definition?

Reply:
Yes, a cstring ends with a `\0` but the point is that the utf8 encoded content here may contain a `\0` byte somewhere in the middle of the string. That is uncommon but possible and might lead to weird/silent bugs down the line. Better to fix it right here. Please use [`initWithBytes`](https://developer.apple.com/documentation/foundation/nsstring/1407339-initwithbytes) with `NSUTF8StringEncoding`. That should do the trick.

Reply:
Done